pkg/adapter/adapter.go

```diff
@@ -31,17 +35,150 @@ type envConfig struct {
+	if err != nil {
+		logger.Desugar().Error("Error creating mongo client", zap.Error(err))
+	}
+	// Get dataSource
+	dataSource := getDataSource(client, env.Database, env.Collection)
+
+	return &mongoDbAdapter{
+		namespace:  env.Namespace,
+		ce:         ceClient,
+		client:     client,
+		datasource: dataSource,
+		source:     env.EventSource,
+		sinkurl:    env.SinkURL,
+		logger:     logger,
+	}
+}
+
+// Returns the appropriate dataSource: database or collection.
+func getDataSource(client *mongo.Client, db string, coll string) *dataSource {
+	if coll != "" {
+		var collsource dataSource = client.Database(db).Collection(coll)
+		return &collsource
+	}
+	var dbsource dataSource = client.Database(db)
+	return &dbsource
+}
+
+// Start creates the watch stream that will watch for dataSource changes.
+func (a *mongoDbAdapter) Start(ctx context.Context) error {
+	// Connect to Client.
+	err := a.client.Connect(ctx)
+	if err != nil {
+		a.logger.Desugar().Error("Error connecting to database", zap.Error(err))
+		return err
+	}
+	defer a.client.Disconnect(ctx)
+
+	// Create a watch stream for either the database or collection.
+	var datasource dataSource = *a.datasource
+	stream, err := datasource.Watch(ctx, mongo.Pipeline{})
+
+	if err != nil {
+		a.logger.Desugar().Error("Error setting up changeStream", zap.Error(err))
+		return err
+	}
+	defer stream.Close(ctx)
+
+	// Watch and process changes.
+	err = a.processChanges(ctx, stream)
+	if err != nil {
+		a.logger.Desugar().Error("Error watching and processing changes", zap.Error(err))
+		return err
+	}
 	return nil
 }
+
+// processChanges processes the new incoming change, creates a cloud event and sends it.
+func (a *mongoDbAdapter) processChanges(ctx context.Context, stream *mongo.ChangeStream) error {
+	// For each new change recorded.
+	for stream.Next(ctx) {
+		var data bson.M
+		if err := stream.Decode(&data); err != nil {
+			a.logger.Desugar().Error("Error decoding the change stream", zap.Error(err))
+		}
+		// Create corresponding event.
+		event, err := makeCloudEvent(data)
+		if err != nil {
+			a.logger.Desugar().Error("Failed to create event", zap.Error(err))
+		}
+
+		// Send event using client.
+		ctx := cloudevents.ContextWithTarget(ctx, a.sinkurl)
+		a.logger.Desugar().Info("", zap.Any("sinkurl", a.sinkurl))
+
+		// Send that Event.
+		if result := a.ce.Send(ctx, *event); cloudevents.IsUndelivered(result) {
+			a.logger.Desugar().Error("Failed to send event", zap.Any("result", result))
+		}
+	}
+	return nil
+}
+
+// makeCloudEvent makes a cloud event out of the change object recevied.
+func makeCloudEvent(data bson.M) (*cloudevents.Event, error) {
+	// Create Event.
+	event := cloudevents.NewEvent(cloudevents.VersionV1)
+
+	// Set cloud event specs and attributes.
+	event.SetID(data["_id"].(bson.M)["_data"].(string))
+	event.SetSource(data["ns"].(bson.M)["db"].(string) + "/" + data["ns"].(bson.M)["coll"].(string))
+	event.SetType(data["operationType"].(string))
```

Comment:
Add to TODO with a link to your issue. Better comments are needed to explain what you're extracting from the bson.